### PR TITLE
syncer: fix table checkpoint panic problem

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/5439

### What is changed and how it work?
1. Add `shardingReSyncUnfinished` to resolve the lowest global checkpoint.
2. Fix the problem that `savedGlobalLastLocation` when shardingResync interruption occurs.
3. Reduce the time we clone locations when comparing.

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?
Possibly.

##### Do you need to update user documentation, design documentation or monitoring documentation?
No.

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/tests/shardddl4_1/run.sh [hunk 1]
```diff
@@ -1032,9 +1032,42 @@ function DM_155 {
...
+	done
+	run_sql_source2 "alter table shardddl1.tb1 change b c int;"
+	for i in $(seq 3001 3100); do
+		run_sql_source1 "insert into shardddl1.tb1(a,c) values($i,$i)"
+	done
+	run_sql_source2 "alter table shardddl1.t_1 change b c int;"
+	for i in $(seq 3101 3200); do
+		run_sql_source1 "insert into shardddl1.tb1(a,c) values($i,$i)"
+	done
+	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml 30
```
Comment (GMHDBJD) [inline dm/tests/shardddl4_1/run.sh:1055]: the diff config file should be updated

Comment (GMHDBJD) [reply dm/tests/shardddl4_1/run.sh:1055]: PANIC_CASE can rename

**Comment by lichunzhu** [comment]
@GMHDBJD @D3Hunter PTAL

**Comment by lichunzhu** [comment]
/run-all-tests

https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/5878/pipeline/

**Comment by lichunzhu** [comment]
/hold
adding some integration tests

**Comment by lichunzhu** [comment]
/run-dm-integration-test

https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/6079/pipeline

**Comment by lichunzhu** [comment]
/run-dm-integration-test

Open API integration test failed

https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/6103/pipeline/204

File: dm/tests/shardddl4_1/run.sh [hunk 2]
```diff
@@ -1041,9 +1053,216 @@ function DM_155 {
...
+	run_sql_with_txn "shardddl1.t_1" 2 231 240 $MYSQL_HOST2 $MYSQL_PORT2 $MYSQL_PASSWORD2
+	check_sync_diff $WORK_DIR $WORK_DIR/diff_config.toml 30
+}
+
+function DM_RESYNC_TXN_INTERRUPT() {
+	run_case RESYNC_TXN_INTERRUPT "double-source-optimistic" \
+		"run_sql_source1 \"create table ${shardddl1}.${tb1} (a int primary key, b int);\"; \
+     run_sql_source2 \"create table ${shardddl1}.t_1 (a int primary key, b int);\"; \
+     run_sql_source2 \"create table ${shardddl1}.${tb2} (a int primary key, b int);\"" \
+		"clean_table" ""
```
Comment (GMHDBJD) [inline dm/tests/shardddl4_1/run.sh:1255]: Better to remove t_1, t_2 here, otherwise test may failed if we add more cases later.

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Comment by lichunzhu** [comment]
@D3Hunter @lance6716 PTAL

File: dm/syncer/syncer.go [hunk 3]
```diff
@@ -1928,56 +1925,41 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
-					// save current location as the next time redirect position
-					shardingReSyncUnfinished = append(shardingReSyncUnfinished, unfinishedShardingResync)
+					shardingReSyncCh <- resync
 				}
+			}
+		})
+		// fetch from sharding resync channel if needed, and redirect global
+		// stream to current binlog position recorded by ShardingReSync
+		if len(shardingReSyncCh) > 0 && shardingReSync == nil {
+			if shardingReSync == nil {
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:1951]: this is not needed, always true.

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

**Review by D3Hunter — APPROVED** [review]

File: dm/syncer/syncer.go [hunk 4]
```diff
@@ -2274,11 +2302,7 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 
 				// only need compare binlog position?
 				lastLocation = shardingReSync.currLocation
-				expectedCmp := 0
-				if s.cfg.ShardMode == config.ShardOptimistic {
-					expectedCmp = 1
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/syncer/syncer.go:2279]: In the past, we didn't strictly demand redirect at the specified event. When we redirect at `currentLocation`, this event is fully ignored before. So we need to compare curr > latest to make sure all this event is replayed.

**Pull request merged**